subprojects/test-runner/service/src/main/kotlin/com/avito/runner/service/worker/DeviceWorker.kt

```diff
@@ -136,62 +123,62 @@ internal class DeviceWorker(
+                    device = device,
+                    targetPackage = action.targetPackage,
                     test = action.test,
-                    result = Failed.InfrastructureError.Unexpected(
-                        error = RuntimeException("Unexpected infrastructure error", t),
+                    executionNumber = action.executionNumber
+                )
+
+                device.runIsolatedTest(
+                    action = action,
+                    outputDir = outputDirectory
+                )
+            } catch (t: Throwable) {
+                val now = timeProvider.nowInMillis()
+
+                DeviceTestCaseRun(
+                    testCaseRun = TestCaseRun(
+                        test = action.test,
+                        result = Failed.InfrastructureError.Unexpected(
+                            error = RuntimeException("Unexpected infrastructure error", t),
+                        ),
+                        timestampStartedMilliseconds = now,
+                        timestampCompletedMilliseconds = now
                     ),
-                    timestampStartedMilliseconds = now,
-                    timestampCompletedMilliseconds = now
-                ),
-                device = device.getData()
-            )
-        }
+                    device = device.getData()
+                )
+            }
+        }.flatMap { deviceTestCaseRun ->
 
-        val reportFileProvider = TestArtifactsProviderFactory.createForAdbAccess(
-            appUnderTestPackage = action.targetPackage,
-            className = action.test.className,
-            methodName = action.test.methodName
-        )
+            val reportFileProvider = TestArtifactsProviderFactory.createForAdbAccess(
+                appUnderTestPackage = action.targetPackage,
+                className = action.test.className,
+                methodName = action.test.methodName
+            )
 
-        val testArtifactsDir = reportFileProvider.provideReportDir()
+            val testArtifactsDir = reportFileProvider.provideReportDir()
 
-        testListener.finished(
-            device = device,
-            test = action.test,
-            targetPackage = action.targetPackage,
-            result = deviceTestCaseRun.testCaseRun.result,
-            durationMilliseconds = deviceTestCaseRun.testCaseRun.durationMilliseconds,
-            executionNumber = action.executionNumber,
-            testArtifactsDir = testArtifactsDir
-        )
+            testListener.finished(
+                device = device,
+                test = action.test,
+                targetPackage = action.targetPackage,
+                result = deviceTestCaseRun.testCaseRun.result,
+                durationMilliseconds = deviceTestCaseRun.testCaseRun.durationMilliseconds,
+                executionNumber = action.executionNumber,
+                testArtifactsDir = testArtifactsDir
+            )
 
-        testArtifactsDir.map { dir ->
```

Comment:
I found two bugs that:
- If `testArtifactsDir` was Failed then map `would` never be called
- After onDeviceDied we must end the coroutine with `return` invocation.